### PR TITLE
Fixing node readdir #7487

### PR DESCRIPTION
**Pull request by AlexanderOMara**
Based on the work @kripken's started. I'm new to emscripten codebase, so let me know if anything isn't correct.

### CODE REVIEW

**Comment by kripken** [comment]
Thanks @AlexanderOMara! Looks like a good start. I'd recommend focusing on the failing tests first. Let me know if you get stuck on something and could use help.

**Comment by AlexanderOMara** [comment]
I think the failing tests are from something in the commits before these changes, and possible even before the raw branch's changes. When I merged the incoming branch with the changes, I wasn't getting the error locally. Waiting for #10170 CI to confirm.

**Comment by AlexanderOMara** [comment]
Well, that fixed `build-upstream-linux` and `build-upstream-mac` but not everything. Have to see about trying to reproduce the other issues locally.

**Comment by AlexanderOMara** [comment]
Well, that fixed `build-upstream-linux` and `build-upstream-mac` but not everything.

Any advice on how to run the tests to reproduce these issues locally?

**Comment by kripken** [comment]
The `asm*.test_` tests use the fastcomp backend, instead of upstream. To test with that locally, you can do `emsdk install tot-fastcomp` and `emsdk activate tot-fastcomp`.

File: src/library_syscall.js [hunk 1]
```diff
@@ -1037,26 +1037,21 @@ var SyscallsLibrary = {
       var id;
       var type;
       var name = stream.getdents[idx];
-      if (name[0] === '.') { // XXX name? see https://github.com/emscripten-core/emscripten/issues/7487
```
Comment (sbc100) [inline src/library_syscall.js]: I think perhaps this code was supposed to be `if (name == '.' or name == '..')`

File: src/library_noderawfs.js [hunk 2]
```diff
@@ -35,7 +38,7 @@ mergeInto(LibraryManager.library, {
     symlink: function() { fs.symlinkSync.apply(void 0, arguments); },
     rename: function() { fs.renameSync.apply(void 0, arguments); },
     rmdir: function() { fs.rmdirSync.apply(void 0, arguments); },
-    readdir: function() { return fs.readdirSync.apply(void 0, arguments); },
+    readdir: function() { return ['.', '..'].concat(fs.readdirSync.apply(void 0, arguments)); },
```
Comment (sbc100) [inline src/library_noderawfs.js]: Doesn't readdirSync already return '.' and '..' as part of the result?  If not then this makes sense yes.

File: src/library_syscall.js (hunk 1 above)
Comment (AlexanderOMara) [reply src/library_syscall.js]: That could be. But would the `id` variable be `1` in both those cases?

File: src/library_noderawfs.js (hunk 2 above)
Comment (AlexanderOMara) [reply src/library_noderawfs.js]: No `readdirSync` actually does not include `'.'` and `'..'`

**Comment by AlexanderOMara** [comment]
Some of the failed tests appear to be related to `test_getdents64_special_cases` however something seems to be wrong about this test.

According to this comment, and the expected output, it seems to think that a filename longer than 255 characters would be truncated when created.

https://github.com/emscripten-core/emscripten/blob/5187ea263120ff57424a184c1bc73259e24dc9a9/tests/fs/test_getdents64_special_cases.cpp#L54-L55

That doesn't seem to be the case. In Node I get a `ENAMETOOLONG` error if a path component is larger than `255` characters, and my little test program also fails to create any file.

```
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>

void rep(char * buffer, char c, int l) {
	for (int i = 0; i < l; i++) {
		buffer[i] = c;
	}
	buffer[l] = '\0';
}

int main() {
	char buffer[1024];
	rep(buffer, '1', 300);
	open(buffer, O_CREAT);
}
```

Should I be truncating the path components in the `library_noderawfs.js` `open` implementation? Or is the test wrong?

**Comment by AlexanderOMara** [comment]
@kripken Any advice on how to proceed with this test?

**Comment by kripken** [comment]
@AlexanderOMara good question. I'm not sure why the test assumes truncation there. Doing some searching, I see stuff like https://www.npmjs.com/package/sanitize-filename which mentions it truncates names to a safe 255 characters, so I guess there's some historical reason? It might be safest to do the same here then.

**Comment by AlexanderOMara** [comment]
@kripken I think I've got all the functional edge-cases taken care of, but I'm still getting some strange errors in the CI I don't understand.

In this pull request, I'm getting some error in the python files on some tests:

> IndexError: list assignment index out of range

Alternately for pull request #10170 (for the incoming branch) I'm getting some other errors on some of the tests:

> AssertionError: 'shared:WARNING: LLVM version appears inco[36 chars]")\n' != ''
> \- shared:WARNING: LLVM version appears incorrect (seeing "11.0", expected "10.0")

Any idea what's going on here?

**Comment by kripken** [comment]
@AlexanderOMara 

For the binaryen error, I think #10244 will fix it. For the LLVM version number, we've changed to use master in emscripten, instead of incoming, and I think pulling latest master to your branch will fix that error (we updated expectations to 11 there).

**Comment by AlexanderOMara** [comment]
@kripken Yep, with the master branch all the tests pass: https://github.com/emscripten-core/emscripten/pull/10245

**Comment by sbc100** [comment]
Bump.. is this still needed or did it land in some other form already?

**Comment by thomasballinger** [comment]
@sbc100 this is still needed. I tried rebasing and updating a test file and it looks good, all tests which pass on main now pass at https://github.com/emscripten-core/emscripten/pull/15167.

**Comment by AlexanderOMara** [comment]
I think so, yeah.
